Extends AArch64 branch protection support to include GCS

CrooseGit opened the pull request:
Extends existing support for AArch64 branch protection to include support for [Guarded Control Stacks](https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/arm-a-profile-architecture-2022#guarded-control-stack-gcs:~:text=Extraction%20or%20tracking.-,Guarded%20Control%20Stack%20(GCS),-With%20the%202022).

CrooseGit committed:
Adds AArch64 GCS support

- Adds option to rustc config to enable GCS
- Passes `guarded-control-stack` flag to llvm if enabled

CrooseGit committed:
Extends branch protection tests to include GCS

davidtwco commented:
I've reviewed this internally so will reroll. GCS is exposed in Clang through their `-mbranch-protection` flag, so this just mirrors that like the rest of this flag.

r? compiler

Urgau reviewed (approved):
Seems pretty straightforward, and looks consistent with clang.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146338/beeb8e3af54295ba494c250e84ecda4c2c5d85ff..b50c4dbf6505d8b2ee488a0bb56f7fdb27c4eae0)*

nnethercote reviewed (commented):
Thanks, @Urgau!

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146338/beeb8e3af54295ba494c250e84ecda4c2c5d85ff..b50c4dbf6505d8b2ee488a0bb56f7fdb27c4eae0)*

CrooseGit commented on tests/ui/invalid-compile-flags/branch-protection-missing-pac-ret.BADFLAGS.stderr:
```diff
@@ -1,2 +1,2 @@
-error: incorrect value `leaf` for unstable option `branch-protection` - a `,` separated combination of `bti`, `pac-ret`, followed by a combination of `pc`, `b-key`, or `leaf` was expected
+error: incorrect value `leaf` for unstable option `branch-protection` - a `,` separated combination of `bti`, `gcs`, `pac-ret`, followed by a combination of `pc`, `b-key`, or `leaf` was expected
```
Ah yes, thank you. I have now fixed it.

Urgau commented:
Does your local `rustc` uses LLVM 19 (like the job)? Or are you building LLVM locally/with `download-ci-llvm`?

davidtwco commented:
Checked locally and it's the LLVM version that is causing this, just add a `//@ min-llvm-version: 21` directive only for the GCS revision (or `//@ min-llvm-version: 20` if it works with that version).

CrooseGit committed:
Changes some aarch64 CIs g++ install & ubuntu ver.

GCS support was added to GCC in version 15, thus the rmake test for this
patch requires GCC15
Similarly, the ubuntu version is updated so the newer clang version is
available, and/or GCC15 is the default.

Urgau commented:
Let's do a try build of all the `aarch64` jobs.

@bors try jobs=`*aarch64*`

Urgau commented on src/ci/docker/host-aarch64/aarch64-gnu-debug/Dockerfile:
We probably need to update the base version of ubuntu to `24.10` to get a new enough `clang>=18` version.

CrooseGit commented:
@bors try jobs=aarch64

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17830315563/job/50693439134) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/50693439134) [(plain)](https://github.com/rust-lang/rust/commit/b984bb58970de3774ee7a30f483dfb6a7058a22b/checks/50693439134/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  libclang-common-17-dev libclang-common-18-dev libclang-rt-16-dev
  libclang-rt-17-dev libclang-rt-18-dev libclang1-16t64 libclang1-17t64
  libcolord2 libcups2t64 libdatrie1 libdav1d7 libdbusmenu-glib4
  libdbusmenu-gtk3-4 libdconf1 libde265-0 libdeflate0 libenchant-2-2 libepoxy0
  libfbclient2 libffi-dev libfftw3-double3 libfontenc1 libgav1-1 libgc1 libgd3
  libgdk-pixbuf-2.0-0 libgdk-pixbuf2.0-bin libgdk-pixbuf2.0-common libglvnd0
  libgraphite2-3 libgtk-3-0t64 libgtk-3-bin libgtk-3-common libharfbuzz0b
  libhashkit2t64 libheif-plugin-aomdec libheif-plugin-libde265 libheif1
  libhunspell-1.7-0 libice6 libicu-dev libimagequant0 libintl-perl
  libintl-xs-perl libjbig0 libjpeg-turbo8 libjpeg8 liblcms2-2 liblerc4
  liblldb-16t64 liblldb-17t64 liblqr-1-0 liblttng-ust-common1t64
  liblttng-ust-ctl5t64 liblttng-ust1t64 libmagickcore-6.q16-7t64
  libmagickwand-6.q16-7t64 libmemcached11t64 libmodule-find-perl
  libmodule-scandeps-perl libncurses-dev libnorm1t64 libobjc-13-dev libobjc4
  libogg0 libopenjp2-7 libpango-1.0-0 libpangocairo-1.0-0 libpangoft2-1.0-0
  libpcre2-16-0 libpcre2-32-0 libpcre2-dev libpcre2-posix3 libpfm4
  libpgm-5.3-0t64 libpixman-1-0 libproc-processtable-perl libqdbm14t64
  librabbitmq4 librav1e0 libraw23t64 librsvg2-2 librsvg2-common libsharpyuv0
  libsm6 libsnappy1v5 libsort-naturally-perl libsvtav1enc1d1 libsybdb5 libtdb1
  libterm-readkey-perl libthai-data libthai0 libtidy5deb1 libtiff6 libtommath1
  libvorbis0a libvorbisfile3 libvulkan1 libwayland-client0 libwayland-cursor0
  libwayland-egl1 libwebp7 libwebpdemux2 libwebpmux3 libxaw7 libxcb-glx0
  libxcb-render0 libxcb-shm0 libxcomposite1 libxcursor1 libxdamage1 libxfixes3
  libxfont2 libxinerama1 libxkbfile1 libxml2-dev libxmu6 libxpm4 libxrandr2
---
Removing libmagickwand-6.q16-7t64:arm64 (8:6.9.12.98+dfsg1-5.2build2) ...
Removing libmagickcore-6.q16-7t64:arm64 (8:6.9.12.98+dfsg1-5.2build2) ...
Removing imagemagick-6-common (8:6.9.12.98+dfsg1-5.2build2) ...
Removing libgd3:arm64 (2.3.3-13+ubuntu24.04.1+deb.sury.org+1) ...
Removing libavif16:arm64 (1.0.4-1ubuntu3) ...
Removing libgav1-1:arm64 (0.18.0-1build3) ...
Removing libabsl20220623t64:arm64 (20220623.1-3.1ubuntu3.2) ...
Removing libaspell15:arm64 (0.60.8.1-1build1) ...
Removing libatspi2.0-0t64:arm64 (2.52.0-1build1) ...
Removing libcups2t64:arm64 (2.4.7-1.2ubuntu7.4) ...
Removing libavahi-client3:arm64 (0.8-13ubuntu6) ...
Removing libavahi-common3:arm64 (0.8-13ubuntu6) ...
---
Removing libimagequant0:arm64 (2.18.0-1build1) ...
Removing libintl-xs-perl (1.33-1build3) ...
Removing libintl-perl (1.33-1build3) ...
Removing libraw23t64:arm64 (0.21.2-2.1ubuntu0.24.04.1) ...
Removing libyuv0:arm64 (0.0~git202401110.af6ac82-1) ...
Removing liblcms2-2:arm64 (2.14-2build1) ...
Removing liblldb-16t64 (1:16.0.6-23ubuntu4) ...
Removing liblldb-17t64 (1:17.0.6-9ubuntu1) ...
Removing liblqr-1-0:arm64 (0.4.2-2.1build2) ...
Removing liblttng-ust1t64:arm64 (2.13.7-1.1ubuntu2) ...
---
Removing libwebpmux3:arm64 (1.3.2-0.4build3) ...
Removing libwebpdemux2:arm64 (1.3.2-0.4build3) ...
Removing libsnappy1v5:arm64 (1.1.10-1build1) ...
Removing libsort-naturally-perl (1.03-4) ...
Removing libsvtav1enc1d1:arm64 (1.7.0+dfsg-2build1) ...
Removing libtdb1:arm64 (1.4.10-1build1) ...
Removing libterm-readkey-perl (2.38-2build4) ...
Removing libthai-data (0.1.29-2build1) ...
Removing libtidy5deb1:arm64 (2:5.6.0-11ubuntu2) ...
Removing libtommath1:arm64 (1.2.1-2build1) ...
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17853071492/job/50765841634) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/50765841634) [(plain)](https://github.com/rust-lang/rust/commit/d67bd7380dbc1b6997a4aede41761cbdafb213e7/checks/50765841634/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
..........................................         (142/142)

======== tests/rustdoc-gui/docblock-table-overflow.goml ========

`tests/rustdoc-gui/docblock-table-overflow.goml` docblock-table-overflow output:
Waiting failed: 30000ms exceeded
stack: TimeoutError: Waiting failed: 30000ms exceeded
    at new WaitTask (/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/node_modules/puppeteer-core/lib/cjs/puppeteer/common/WaitTask.js:50:34)
    at IsolatedWorld.waitForFunction (/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/node_modules/puppeteer-core/lib/cjs/puppeteer/api/Realm.js:25:26)
    at CdpFrame.waitForFunction (/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/node_modules/puppeteer-core/lib/cjs/puppeteer/api/Frame.js:561:43)
    at CdpFrame.<anonymous> (/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/node_modules/puppeteer-core/lib/cjs/puppeteer/util/decorators.js:98:27)
    at CdpPage.waitForFunction (/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/node_modules/puppeteer-core/lib/cjs/puppeteer/api/Page.js:1366:37)
    at runAllCommands (/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/node_modules/browser-ui-test/src/index.js:418:28)
    at async innerRunTestCode (/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/node_modules/browser-ui-test/src/index.js:696:21)
    at async innerRunTests (/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/node_modules/browser-ui-test/src/index.js:633:17)
    at async runTests (/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/node_modules/browser-ui-test/src/index.js:824:27)


======== tests/rustdoc-gui/search-result-display.goml ========

[WARNING] `tests/rustdoc-gui/search-result-display.goml` line 39: Delta is 0 for "x", maybe try to use `compare-elements-position` instead?
```

</details>

davidtwco commented:
cc @rust-lang/infra for the CI image changes

rust-log-analyzer commented:
The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17864542971/job/50803669583) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/50803669583) [(plain)](https://github.com/rust-lang/rust/commit/c3587c993a8edecc8a3482a86fe9cda2f77fe8dd/checks/50803669583/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:start] test::Compiletest { test_compiler: Compiler { stage: 2, host: aarch64-apple-darwin, forced_compiler: false }, target: aarch64-apple-darwin, mode: "assembly", suite: "assembly-llvm", path: "tests/assembly-llvm", compare_mode: None }
##[group]Testing stage2 with compiletest suite=assembly-llvm mode=assembly (aarch64-apple-darwin)

running 570 tests
test [assembly] tests/assembly-llvm/aarch64-pointer-auth.rs#GCS ... ignored, ignored when the target vendor is Apple ((XCode version needs updating))
test [assembly] tests/assembly-llvm/aarch64-pointer-auth.rs#PACRET ... ignored, ignored when the target vendor is Apple ((XCode version needs updating))
test [assembly] tests/assembly-llvm/aarch64-pointer-auth.rs#PAUTHLR ... ignored, ignored when the target vendor is Apple ((XCode version needs updating))
test [assembly] tests/assembly-llvm/aarch64-pointer-auth.rs#PAUTHLR_NOP ... ignored, ignored when the target vendor is Apple ((XCode version needs updating))
test [assembly] tests/assembly-llvm/aarch64-xray.rs#aarch64-linux ... ignored, only executed when the target is aarch64-unknown-linux-gnu
test [assembly] tests/assembly-llvm/align_offset.rs ... ignored, only executed when the architecture is x86_64
test [assembly] tests/assembly-llvm/aarch64-xray.rs#aarch64-darwin ... ok
test [assembly] tests/assembly-llvm/asm/aarch64-outline-atomics.rs ... ignored, only executed when the operating system is linux
test [assembly] tests/assembly-llvm/asm/aarch64-el2vmsa.rs ... ok
---
---- [run-make] tests/run-make/pointer-auth-link-with-c stdout ----

error: rmake recipe failed to complete
status: exit status: 1
command: cd "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/pointer-auth-link-with-c/rmake_out" && env -u RUSTFLAGS -u __RUSTC_DEBUG_ASSERTIONS_ENABLED -u __STD_DEBUG_ASSERTIONS_ENABLED AR="ar" BUILD_ROOT="/Users/runner/work/rust/rust/build/aarch64-apple-darwin" CC="/Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx -mmacosx-version-min=11 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk" CXX="/Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx -mmacosx-version-min=11 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk -stdlib=libc++" DYLD_LIBRARY_PATH="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/bootstrap-tools/aarch64-apple-darwin/release/deps:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0/lib/rustlib/aarch64-apple-darwin/lib" HOST_RUSTC_DYLIB_PATH="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib" LD_LIB_PATH_ENVVAR="DYLD_LIBRARY_PATH" LLVM_BIN_DIR="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/ci-llvm/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard cgdata codegen codegentypes core coroutines coverage csky cskyasmparser cskycodegen cskydesc cskydisassembler cskyinfo debuginfobtf debuginfocodeview debuginfodwarf debuginfodwarflowlevel debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarfcfichecker dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontendatomic frontenddirective frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld sandboxir scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser telemetry textapi textapibinaryreader transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/ci-llvm/bin/FileCheck" NODE="/opt/homebrew/bin/node" PYTHON="/usr/bin/python3" RUSTC="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustc" RUSTDOC="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustdoc" SOURCE_ROOT="/Users/runner/work/rust/rust" TARGET="aarch64-apple-darwin" TARGET_EXE_DYLIB_PATH="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/lib" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/pointer-auth-link-with-c/rmake"
stdout: none
--- stderr -------------------------------
command failed at line 23
Command { cmd: "/Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=11" "-isysroot" "/Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk" "-v" "-c" "-o" "test" "test.c" "-mbranch-protection=bti+gcs+pac-ret+leaf", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "/Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang", defused: true, armed_location: Location { file: "/Users/runner/work/rust/rust/tests/run-make/pointer-auth-link-with-c/rmake.rs", line: 18, column: 5 } }, already_executed: true }
output status: `exit status: 1`
=== STDOUT ===



=== STDERR ===
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-macosx
Thread model: posix
InstalledDir: /Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
clang: error: unsupported argument 'gcs' to option '-mbranch-protection='
------------------------------------------

---- [run-make] tests/run-make/pointer-auth-link-with-c stdout end ----

failures:
    [run-make] tests/run-make/pointer-auth-link-with-c

test result: FAILED. 324 passed; 1 failed; 106 ignored; 0 measured; 0 filtered out; finished in 106.94s
```

</details>

jieyouxu commented on src/ci/docker/host-aarch64/aarch64-gnu-debug/Dockerfile:
```diff
@@ -1,8 +1,8 @@
-FROM ubuntu:22.04
+FROM ubuntu:25.04
```
Question: doesn't this mean that we no longer test if we can build against Ubuntu 22?

jieyouxu commented on src/ci/docker/host-aarch64/aarch64-gnu-debug/Dockerfile:
```diff
@@ -1,8 +1,8 @@
 ARG DEBIAN_FRONTEND=noninteractive
 RUN apt-get update && apt-get install -y --no-install-recommends \
-  g++ \
+  g++-15 \
```
Question: ... and for that matter, is this a higher g++ baseline? (I'm not super sure.)

bors commented:
:hourglass: Testing commit 08020def99d2851af0dabde12cc6d203017fa72c with merge 15283f6fe95e5b604273d13a428bab5fc0788f5a...
<!-- homu: {"type":"BuildStarted","head_sha":"08020def99d2851af0dabde12cc6d203017fa72c","merge_sha":"15283f6fe95e5b604273d13a428bab5fc0788f5a"} -->

rust-timer commented:
Finished benchmarking commit (15283f6fe95e5b604273d13a428bab5fc0788f5a): [comparison URL](https://perf.rust-lang.org/compare.html?start=e9385f9eea0221ef295a188d49d16f8f5189abf1&end=15283f6fe95e5b604273d13a428bab5fc0788f5a&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=e9385f9eea0221ef295a188d49d16f8f5189abf1&end=15283f6fe95e5b604273d13a428bab5fc0788f5a&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.2%, 0.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e9385f9eea0221ef295a188d49d16f8f5189abf1&end=15283f6fe95e5b604273d13a428bab5fc0788f5a&stat=max-rss)
<details>
<summary>Results (primary -0.7%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.1%  | [2.1%, 2.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.5% | [-3.5%, -3.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.7% | [-3.5%, 2.1%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e9385f9eea0221ef295a188d49d16f8f5189abf1&end=15283f6fe95e5b604273d13a428bab5fc0788f5a&stat=cycles:u)
<details>
<summary>Results (secondary -3.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.1% | [-3.5%, -2.7%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=e9385f9eea0221ef295a188d49d16f8f5189abf1&end=15283f6fe95e5b604273d13a428bab5fc0788f5a&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 471.559s -> 471.213s (-0.07%)
**Artifact size**: 387.91 MiB -> 387.83 MiB (-0.02%)
